### PR TITLE
NIFI-5514: Fixed bugs in MergeRecord around minimum thresholds not be…

### PR DESCRIPTION
**Pull request by markap14**
…ing honored and validation not being performed to ensure that minimum threshold is smaller than max threshold (would previously allow min record = 100, max records = 2 as a valid configuration)

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by bdesert** [comment]
Reviewing...

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MergeRecord.java [hunk 1]
```diff
@@ -304,13 +336,25 @@ public void onTrigger(final ProcessContext context, final ProcessSessionFactory
             session.commit();
         }
 
+        // If there is no more data queued up, complete any bin that meets our minimum threshold
+        int completedBins = 0;
+        final QueueSize queueSize = session.getQueueSize();
```
Comment (bdesert) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MergeRecord.java]: @markap14 
Test for MinRecords is successful, but actual flow doesn't work as expected and still doesn't respect Min Records setting.
There is my set up:
image

image

The problem is in different behavior of MockProcessSession and StandardProcessSession.
MockProcessSession.getQueueSize() will return 0, after session.get(...)
StandardProcessSession.getQueueSize() will return same number as before session.get(...), regardless flow files have been polled or not.
As a result, this condition will block from actually emitting FF when min requirements are reached.

I would recommend to change this condition to:
if (flowFiles.size() != 0) {...}

In parallel, we probably need to open a JIRA for inconsistency between MockProcessSession and StandardProcessSession.

Comment (markap14) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MergeRecord.java]: @bdesert that's a great catch! I have pushed a new commit that updates the code like you suggested, to just check the size of the flowFiles list. I also created NIFI-5626 to address the inconsistency between MockProcessSession and StandardProcessSession.

**Comment by bdesert** [comment]
+1 LGTM.
Test case updated.
Live test on local env (up to date) succeeded. Works as expected.
Travis is failing for JP only (US and FR are OK). Can be merged.

**Comment by pvillard31** [comment]
+1, merging to master, thanks @markap14

**Pull request closed without merging**